Generate doc for 3 doxygen versions and provide a script to extract it.

maxGimeno opened the pull request:
Fix CGAL/cgal#1800.

This PR changes the container so it generates the documentation for three different versions of doxygen :
- the master of the cgal fork
- the branch release_1_8_13_patched from the cgal fork
- the master of doxygen

It also provides a script to extract this documentation. The script generate_documentation.sh must be called from the directory in which the doc will be extracted. Its first argument is the path to the Dockerfile of this container, and its second argument is the path the the CGAL directory from which the documentation will be built.

Depends on https://github.com/CGAL/cgal/pull/4448

maxGimeno committed:
Generate doc for 3 doxygen versions and provide a script to extract it.

lrineau commented on Dockerfile:
```diff
@@ -1,24 +1,53 @@
+ #   python2-pyquery \
+ #   texlive-bin \
+ #   && \
+ #   pacman -Scc --noconfirm
```
You know what I think about committing commented code: just remove the commented lines!

maxGimeno committed:
Remove commented code

lrineau commented on README.md:
```diff
@@ -1,18 +1,18 @@
 ==================
 
-This Docker image can be used to build the documentation of CGAL.
+These Docker images can be used to build the documentation of CGAL.
```
There is still only one image, even if one image allows to build with three different versions of Doxygen.

lrineau commented on Dockerfile:
```diff
@@ -1,24 +1,42 @@
+     texlive-binaries \
+  && apt-get clean -y \
 
 USER makepkg
```
That line probably triggers an error, or at least a warning. I doubt the `ubuntu` image has a `makepkg` user.

lrineau commented on README.md:
```diff
@@ -1,18 +1,18 @@
-    docker run -v /path/to/cgal:/mnt/cgal \
-               cgal-doxygen
+    docker run -v /path/to/cgal:/mnt/cgal -t -i \
+               cgal-doxygen /bin/bash/docker_entrypoint.sh cgal_build_documentation
```
> `/bin/bash/docker_entrypoint.sh`

... there must be a space after `bash`.

lrineau commented on generate_documentation.sh:
```diff
@@ -0,0 +1,27 @@
+#docker run -v "$PATH_TO_CGAL":/mnt/cgal -t cgal_doc /docker_entrypoint.sh cgal_build_documentation
+#run second time to link
+#docker run -v "$PATH_TO_CGAL":/mnt/cgal -t cgal_doc /docker_entrypoint.sh cgal_build_documentation
+#create output directories
```
I find strange that the script does not launch a container at all.

maxGimeno committed:
Fixes

sloriot commented on Dockerfile:
```diff
@@ -1,24 +1,41 @@
-    cd doxygen && ./configure && \
+
+RUN git clone https://github.com/CGAL/doxygen.git cgal_master && \
+    cd cgal_master && \
```
you should use the branch `release_1_8_4_patched` as master might change

sloriot commented on Dockerfile:
```diff
@@ -1,24 +1,41 @@
+    cd build && \
+    cmake .. && \
+    make
+
```
Not a big deal but you clone 3 times a repo to access 3 of its branches.

sloriot commented on generate_documentation.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
@lrineau I let you review this one as I don't know which scripts it replaces.

lrineau replied:
Actually, I see that this docker image was first build by Phillip, but it seems it is not used in our daily testing of the CGAL documentation.

We could try to run that docker image in parallel to the other process, evaluate the results, and eventually switch to the Docker-based process.

But first Maxime needs to modify the code according to the comments you made two weeks ago.

lrineau replied:
The current script that generates the Doxygen test results is [`Scripts/developer_scripts/run_doxygen_testsuite`](https://github.com/CGAL/cgal/blob/760076510cf8c7514f85d2dceb6123fc557479b7/Scripts/developer_scripts/run_doxygen_testsuite).

lrineau commented:
@maxGimeno The feedback from Sébastien was left unanswered.

maxGimeno committed:
Use 1_8_4 instead of master and don't clone 3 times.

lrineau commented on generate_documentation.sh:
```diff
@@ -0,0 +1,27 @@
+#generate documentation
+docker run -v "$PATH_TO_CGAL":/mnt/cgal -t cgal_doc /docker_entrypoint.sh cgal_build_documentation
+#run second time to link
+docker run -v "$PATH_TO_CGAL":/mnt/cgal -t cgal_doc /docker_entrypoint.sh cgal_build_documentation
```
Those runs of the container cannot work: between the two runs, the content of all the filesystem of the previous run will be lost! Containers are one shot and self-contained (but in Docker volumes).

If one needs to run twice the `make doc` step, then let's do in in the script ran in the container.

lrineau commented on generate_documentation.sh:
```diff
@@ -0,0 +1,27 @@
+  docker cp "$DOCKER_ID":/"cgal_build_$dir/doc_output" "./$dir"
+  docker cp "$DOCKER_ID":"/cgal_build_$dir/doc_log" "./$dir"
+  mkdir -p "$dir/testsuite_out"
+  python "$PATH_TO_CGAL/Documentation/doc/scripts/testsuite.py"  --doc-log-dir "$dir/doc_log" --output-dir "$dir/doc_output" --cgal-version 1 --publish "$PWD/$dir/testsuite_out"
```
@maxGimeno Why is not the script `Documentation/doc/scripts/testsuite.py` run from inside the container?

maxGimeno replied:
isn't this script used to upload the content of the documentation build ? can it be done from inside a container ?

lrineau replied:
One can attach "volumes" to containers, with the `-v` option of `docker run`. The volumes are the way for a container to access a part of the host filesystem. Those volumes can be read-only but they can also be read-writable.

lrineau replied:
That's how the testsuite container works: there is a read-only volume that contains the input (the tarballs of the releases), and a read-write volume that is dedicated to the output of the test results.

maxGimeno committed:
Fixes :

- Call make doc twice in the docker_entrypoint.sh instead of doing so in the generate_documentation.sh
- Build the doxygen versions in /doxygen instead of tmp/makepg so it is not discarded at the end of the build.

maxGimeno commented:
@sloriot Do you have other remarks ? Are we good ?

lrineau commented:
Maxime, what does it mean "are we good"? What is you proposal to replace the script `Scripts/developer_scripts/run_doxygen_testsuite` that I mentioned in https://github.com/CGAL/cgal-documentation-dockerfile/pull/1#discussion_r99619024?

maxGimeno commented:
You said
> We could try to run that docker image in parallel to the other process, evaluate the results, and eventually switch to the Docker-based process.

Does it mean I should modify the run_doxygen_testsuite script to call docker ?

lrineau commented:
This PR should use the scripts that were created/modified by the PR CGAL/cgal#1852.

lrineau commented:
@maxGimeno This PR should be revived, and upgraded to the new code CGAL/cgal#1852.

We discussed with @sloriot, and our objective is to start testing the three versions of Doxygen before the start of the release cycle of 4.13. That means before mid-January 2018.

lrineau commented:
This task adds two things:
  1. The documentation is compiled in a Docker container, instead of doing it in our usual user directories.
  2. Test three versions of Doxygen instead of the current two:
     - test with Doxygen 1.8.4 patched by CGAL
     - test with Doxygen 1.8.13 patched by CGAL (the one that is added),
     - test with Doxygen `master`.

The official switch to 1.8.13-patched will be during the release cycle of 4.13, when the tests are good.

maxGimeno committed:
Use the recent comparison script from cgal

maxGimeno commented:
I think it is ready now. It basically performs the process to generate the 3 docs but in a docker container.

maxGimeno commented:
@lrineau does my last commit solve your changes request ?

maxGimeno commented:
This PR is useless now, isn't it ?

lrineau commented:
No. For the moment, the Doxygen testsuite is run directly by a cron job. It should be run in a container.

maxGimeno committed:
replace python by 3.5 and move publish dir in doc/scripts instead of / so it can be get fetched back afterwards

maxGimeno committed:
install python 3.5

lrineau commented:
Does this PR work? Should we opt for this Docker images, to build our documentation, instead of our current testsuite process (with cron jobs)?

maxGimeno commented:
Last time I checked, it worked. I used it several times to test fixes in the doc

lrineau commented:
I think we should stop testing three versions of Doxygen. Version 1.8.4 is no longer used with current supported CGAL version (4.14, 5.0, 5.1).

maxGimeno committed:
Replace 1.8.4 by 1.8.14

lrineau commented:
Well, you replaced 1.8.4 with 1.8.14. I suggest you use the official release 1.8.18 instead of 1.8.14.

maxGimeno committed:
replace 1.8.14 with 1.8.18

maxGimeno committed:
Fixes

lrineau commented on Dockerfile:
```diff
@@ -6,10 +6,11 @@ RUN  apt-get update \
      flex \
      graphviz \
      python3 \
+     python2 \
```
Really?

maxGimeno replied:
Sadly yes, really. Doxygen is unable to process \ref without this.

bo0ts replied:
"Progress is not in enhancing what is, but in advancing toward what will be." Khalil Gibran

Glad someone is watching out for it.

maxGimeno committed:
Remove python2, the problem is fixed now in CGAL

maxGimeno committed:
Remove 1.8.4 and test 1.9.1

lrineau commented on docker_entrypoint.sh:
```diff
@@ -4,9 +4,9 @@ set -e
-    mkdir -p $PWD/doc_1_8_13
-    mkdir -p $PWD/doc_1_8_18
-    bash ./process_doc.sh "/doxygen/cgal_1_8_13/doxygen" "/doxygen/cgal_1_8_18/doxygen" /testsuite_out
+    mkdir -p $PWD/doc_1_9_1
+    mkdir -p $PWD/doc_1_9_1
```
That is twice the same directory.

maxGimeno committed:
fix typo

maxGimeno committed:
Fix typo

lrineau commented on Dockerfile:
```diff
@@ -1,24 +1,43 @@
+
+
+
+RUN git clone https://github.com/doxygen/doxygen.git off_dox && \
+    mkdir cgal_1_9_1 && \
+    cd off_dox && \
+    git checkout Release_1_9_1 && \
```
That is not correct. We want a version of Doxygen 1.9.1 that is modified by the same patch as our custom version of Doxygen.

@sloriot Would you mind preparing a branch or a tag `release_1_9_1_patched` in https://github.com/CGAL/doxygen.git?

sloriot replied:
Done. Did not pick the xml fix as it seems no longer needed.

lrineau commented:
In `~/Git/cgal-master` I have a checkout of https://github.com/CGAL/cgal/pull/5847. Then I type:

```shellsession
[lrineau@bonnard]~/Git/cgal-documentation-dockerfile% ./generate_documentation.sh $PWD ~/Git/cgal-master ~/tmp/test-doc
```
and I get:
```
Sending build context to Docker daemon  145.9kB
Step 1/11 : FROM cgal/testsuite-docker:ubuntu
 ---> 78ce820da314
Step 2/11 : MAINTAINER Philipp Moeller <bootsarehax@gmail.com>
 ---> Using cache
 ---> 94285062ffcd
[...]
Step 11/11 : CMD ["cgal_build_documentation"]
 ---> Using cache
 ---> 5ccb57b7934b
Successfully built 5ccb57b7934b
Successfully tagged cgal_documentation:comparison
downloading and building master
fatal: destination path 'doxygen_master' already exists and is not an empty directory.
 clone of doxygen failed
 pull of doxygen failed
 cmake of doxygen failed
done.
[...]
```
Those errors
```
 clone of doxygen failed
 pull of doxygen failed
 cmake of doxygen failed
```
are worrisome, in my opinion. But the script goes on anyway:
```
fatal: destination path 'doxygen_master' already exists and is not an empty directory.
 clone of doxygen failed
 pull of doxygen failed
 cmake of doxygen failed
done.
comparing versions 1_8.13 and 1.9.1
Building reference documentation...
done.
Creating text files for diff....
done.
Building second documentation...
done.
Comparing results...
done.
Adding postprocessing...
done.
Building reference documentation with postprocessing...
done.
CMake Error: The current CMakeCache.txt directory /mnt/cgal/build/CMakeCache.txt is different than the directory /home/lrineau/Git/cgal-master/build where CMakeCache.txt was created. This may result in binaries being created in the wrong place. If you are not sure, reedit the CMakeCache.txt
CMake Error: The source "/mnt/cgal/CMakeLists.txt" does not match the source "/home/lrineau/Git/cgal-master/CMakeLists.txt" used to generate cache.  Re-run cmake with a different source directory.
comparing versions 1_8.13 and master
 test_doxygen_versions with master failed
mv: cannot stat 'build_doc/build_logs': No such file or directory
cat: cgal_version: No such file or directory
NO MASTER
Traceback (most recent call last):
  File "/mnt/cgal/Documentation/doc/scripts/testsuite.py", line 366, in <module>
    main()
  File "/mnt/cgal/Documentation/doc/scripts/testsuite.py", line 219, in main
    d,sum1,sum2,mastersum=write_report(args)
  File "/mnt/cgal/Documentation/doc/scripts/testsuite.py", line 128, in write_report
    os.chdir(args.doc_log_dir2)
FileNotFoundError: [Errno 2] No such file or directory: '/mnt/cgal/Documentation/doc/scripts/doc_1_9_1/doc_log/'
rm: cannot remove './cgal_version': No such file or directory
```

That script has garbled by `/home/lrineau/Git/cgal-master/build` directory, if I read well! :disappointed:

lrineau commented:
> That script has garbled by `/home/lrineau/Git/cgal-master/build` directory, if I read well! disappointed

Actually, my `/home/lrineau/Git/cgal-master/build` was left untouched. The script quitted without touching it.

I have tried again with a cleaned repository (`git clean -x -d`), and it worked well. I will now try to see how we can transition to one way of building the doc to another.

lrineau commented:
For the transition, I think we can modify `Scripts/developer_scripts/run_doxygen_testsuite` in the PR https://github.com/CGAL/cgal/pull/5847, so that is calls `./generate_documentation.sh`. I will do a last check, so see if that works well on the server `cgal.geometryfactory.com` (that still has `/use/bin/python` at a version 2.7).
